Introduce ResponseInstructions for OnionMessage Handling

shaavan opened the pull request:
resolves #2882

Introduce ResponseInstructions for OnionMessage Handling

- Currently, handle_message (or handle_custom_message) only exposes the generated response for an OnionMessage, lacking the necessary reply_path for asynchronous responses.
- This PR introduces a new flow for OnionMessage handling.
- Instead of solely taking the message as input, handle_message now accepts an Optional `responder`, returning a `ResponseInstruction` containing both the response and the reply_path.
- `ResponseInstruction` utilizes different enum variants to indicate how `handle_onion_message_response` should handle the response.
- This enhancement enables exposing the reply_path alongside the response and allows for more complex response mechanisms, such as responding with an added reply_path.

jkczyz commented on lightning/src/onion_message/messenger.rs:
```diff
@@ -243,6 +243,54 @@ impl OnionMessageRecipient {
 
+pub struct Responder<'a, OMH: OnionMessageHandler, T: OnionMessageContents> {
+	messenger: &'a OMH,
+	pub message: T,
```
Let's leave the `message` out of responder and put it directly in `ReceivedOnionMessage::WithReplyPath`. You can add a `&'static str` field for the message type.

shaavan replied:
Sounds reasonable! Addressed in [8526afc](https://github.com/lightningdevkit/rust-lightning/pull/2907/commits/8526afc0989c69249a4fca1f10d03d15dc1a9c15)

Also instead of introducing message type as a separate variable, I used the fact that we enforce response and message to be of the same type at compile time.
Let me know if we should rather have a separate message type variable, and I would love to make that change.

jkczyz commented on lightning/src/onion_message/messenger.rs:
```diff
@@ -243,6 +243,54 @@ impl OnionMessageRecipient {
+    }
+}
+
+pub struct MessageWithId<T: OnionMessageContents> {
+    pub message: T,
+    pub path_id: Option<[u8; 32]>,
+}
```
Let's drop this in favor of using named fields in the enum variant.

```rust
WithoutReplyPath {
    message: T,
    path_id: Option<[u8; 32]>,
}
```

shaavan replied:
Addressed in [8526afc](https://github.com/lightningdevkit/rust-lightning/pull/2907/commits/8526afc0989c69249a4fca1f10d03d15dc1a9c15)=
Thanks!

jkczyz commented on lightning/src/onion_message/messenger.rs:
```diff
@@ -243,6 +243,54 @@ impl OnionMessageRecipient {
+    pub path_id: Option<[u8; 32]>,
+}
+
+pub enum ReceivedOnionMessage<'a, OMH: OnionMessageHandler, T: OnionMessageContents> {
```
Please add docs.

shaavan replied:
Addressed in [8526afc](https://github.com/lightningdevkit/rust-lightning/pull/2907/commits/8526afc0989c69249a4fca1f10d03d15dc1a9c15)

Currently added very rudimentary docs. Let me know if you have some recommendations to expand to docs, and I would love to incorporate them!

jkczyz reviewed (commented):
First pass

shaavan commented on lightning/src/onion_message/messenger.rs:
```diff
@@ -500,7 +548,7 @@ pub trait CustomOnionMessageHandler {
 	///
 	/// The returned [`Self::CustomMessage`], if any, is enqueued to be sent by [`OnionMessenger`].
-	fn handle_custom_message(&self, msg: Self::CustomMessage) -> Option<Self::CustomMessage>;
+	fn handle_custom_message<OMH: OnionMessageHandler>(&self, received_onion_message: &ReceivedOnionMessage<OMH, Self::CustomMessage>);
```
Renamed it in [8526afc](https://github.com/lightningdevkit/rust-lightning/pull/2907/commits/8526afc0989c69249a4fca1f10d03d15dc1a9c15)
Thanks!

shaavan commented on lightning/src/onion_message/messenger.rs:
```diff
@@ -243,6 +243,66 @@ impl OnionMessageRecipient {
+    }
+}
+
+/// A enum to handle received [`OnionMessage`]
+pub enum ReceivedOnionMessage<'a, OMH: OnionMessageHandler, T: OnionMessageContents> {
+	WithReplyPath {
+		message: T,
+		responder: Responder<'a, OMH, T>,
+	},
+	WithoutReplyPath {
+		message: T,
+		path_id: Option<[u8; 32]>,
+	},
+}
+
+impl<'a, OMH: OnionMessageHandler, T: OnionMessageContents> ReceivedOnionMessage<'a, OMH, T> {
+	fn new(messenger: &'a OMH, message: T, reply_path_option: Option<BlindedPath>, path_id: Option<[u8; 32]> ) -> Self {
+		match reply_path_option {
+			Some(reply_path) => {
+				let responder = Responder {
+					messenger,
+					reply_path,
+					path_id,
+					_phantom: std::marker::PhantomData
+				};
+				ReceivedOnionMessage::WithReplyPath {
+					message,
+					responder
+				}
+			}
+			None => ReceivedOnionMessage::WithoutReplyPath {
+				message,
+				path_id
+			}
+		}
+    }
+}
```
Sure!

jkczyz commented on lightning/src/onion_message/messenger.rs:
```diff
@@ -243,6 +243,66 @@ impl OnionMessageRecipient {
 
+/// A struct handling response to an [`OnionMessage`]
+pub struct Responder<'a, OMH: OnionMessageHandler, T: OnionMessageContents> {
+	messenger: &'a OMH,
```
Instead of expanding `OnionMessageHandler`'s definition, could we use a `Fn` or similar trait?

jkczyz commented on lightning/src/onion_message/messenger.rs:
```diff
@@ -243,6 +243,66 @@ impl OnionMessageRecipient {
+	pub path_id: Option<[u8; 32]>,
+	// This phantom Data is used to ensure that we use T in the struct definition
+	// This allow us to ensure at compile time that the received message type and response type will be same
+	_phantom: std::marker::PhantomData<T>
```
Please use `core` instead of `std`.

jkczyz commented on lightning/src/onion_message/messenger.rs:
```diff
@@ -243,6 +243,66 @@ impl OnionMessageRecipient {
+			response, self.reply_path.clone(), format_args!(
+				"when responding to {} onion message with path_id {:02x?}",
+				message_type,
+				self.path_id.clone()
```
Is the `clone` needed?

jkczyz commented on lightning/src/onion_message/messenger.rs:
```diff
@@ -243,6 +243,66 @@ impl OnionMessageRecipient {
+		// Utilising the fact that we ensure at compile time that
+		// received message type, and response type will be same
+		let message_type = T::msg_type(&response);
+        self.messenger.handle_onion_message_response(
```
Use tabs instead of spaces.

jkczyz commented on lightning/src/onion_message/messenger.rs:
```diff
@@ -243,6 +243,66 @@ impl OnionMessageRecipient {
+pub struct Responder<'a, OMH: OnionMessageHandler, T: OnionMessageContents> {
+	messenger: &'a OMH,
+	pub reply_path: BlindedPath,
+	pub path_id: Option<[u8; 32]>,
```
I'm not entirely sure if it is possible, but it would be nice to have some symmetry in the enum variants by having the `path_id` directly in both. Perhaps by storing `FormatArgs`? But you may run into some lifetime issues. Copying the path_id wouldn't be horrible, though.

shaavan commented:
Updated from [pr2907.02](https://github.com/shaavan/rust-lightning/commits/pr2907.02) to [pr2907.03](https://github.com/shaavan/rust-lightning/commits/pr2907.03) ([diff](https://github.com/shaavan/rust-lightning/compare/pr2907.02..pr2907.03)):
Addressed @jkczyz comments

1. Handle fixups
2. Introduce `Fn Traits` in `Responder`

jkczyz commented on lightning/src/onion_message/messenger.rs:
```diff
@@ -250,19 +250,19 @@ pub struct Responder<'a, OMH: OnionMessageHandler, T: OnionMessageContents> {
 }
 
 impl<'a, OMH: OnionMessageHandler, T: OnionMessageContents> Responder<'a, OMH, T> {
     pub fn respond(&self, response: T) {
```
Need a tab here and the closing parenthesis, too.

jkczyz commented on lightning/src/ln/msgs.rs:
```diff
@@ -1590,6 +1592,8 @@ pub trait OnionMessageHandler: EventsProvider {
 	/// Handle an incoming `onion_message` message from the given peer.
 	fn handle_onion_message(&self, peer_node_id: &PublicKey, msg: &OnionMessage);
 
+	fn handle_onion_message_response<T: OnionMessageContents>(&self, response: T, reply_path: BlindedPath, log_suffix: fmt::Arguments);
+
```
Adding this to the trait is no longer necessary.

shaavan replied:
Moved the fn back to it's original position in [2806122](https://github.com/lightningdevkit/rust-lightning/pull/2907/commits/2806122f65d57d3a6a91591aff6e9ccb7590af13)
Thanks!

jkczyz commented on lightning/src/onion_message/messenger.rs:
```diff
@@ -243,6 +244,89 @@ impl OnionMessageRecipient {
+
+impl<F, T> RespondFunction<T> for F
+where
+T: OnionMessageContents,
+F: Fn(T, BlindedPath)
```
Please indent these here and elsewhere.

shaavan replied:
Addressed in [2806122](https://github.com/lightningdevkit/rust-lightning/pull/2907/commits/2806122f65d57d3a6a91591aff6e9ccb7590af13)
Thanks!

jkczyz commented on lightning/src/onion_message/offers.rs:
```diff
@@ -40,7 +43,7 @@ pub trait OffersMessageHandler {
 	///
 	/// [`OnionMessenger`]: crate::onion_message::messenger::OnionMessenger
-	fn handle_message(&self, message: OffersMessage) -> Option<OffersMessage>;
+	fn handle_message<F: Fn(OffersMessage, BlindedPath)>(&self, message: &ReceivedOnionMessage<F, OffersMessage>);
```
Couple things here:
- Take `message` by value as it was. That way handlers can take ownership of the argument and reply asynchronously
- `F` should now be bound by `RespondFunction`.

Likewise for `CustomOnionMessageHandler`.

shaavan replied:
Addressed in 9699624
Thanks!

jkczyz commented on lightning/src/onion_message/messenger.rs:
```diff
@@ -243,6 +244,89 @@ impl OnionMessageRecipient {
+    pub fn respond(&self, response: T) {
+		// Utilising the fact that we ensure at compile time that
+		// received message type, and response type will be same
+		self.messenger_function.respond(response, self.reply_path.clone())
```
We can avoid this `clone` by taking `self` by value.

shaavan replied:
Addressed in [2806122](https://github.com/lightningdevkit/rust-lightning/pull/2907/commits/2806122f65d57d3a6a91591aff6e9ccb7590af13)
Thanks!

jkczyz commented on lightning/src/onion_message/messenger.rs:
```diff
@@ -243,6 +244,89 @@ impl OnionMessageRecipient {
+T: OnionMessageContents
+{
+    pub fn respond(&self, response: T) {
+		// Utilising the fact that we ensure at compile time that
+		// received message type, and response type will be same
```
You can drop this comment as it repeats the comment on `_phantom`.

shaavan replied:
Addressed in [2806122](https://github.com/lightningdevkit/rust-lightning/pull/2907/commits/2806122f65d57d3a6a91591aff6e9ccb7590af13)
Thanks!

jkczyz commented on lightning/src/onion_message/messenger.rs:
```diff
@@ -243,6 +244,89 @@ impl OnionMessageRecipient {
+T: OnionMessageContents
+{
+	messenger_function: R,
+	pub reply_path: BlindedPath,
```
Does this need to be `pub`?

shaavan replied:
I think this was added for the previous version of the PR.

Addressed in [2806122](https://github.com/lightningdevkit/rust-lightning/pull/2907/commits/2806122f65d57d3a6a91591aff6e9ccb7590af13)
Thanks!

jkczyz commented on lightning/src/onion_message/messenger.rs:
```diff
@@ -243,6 +244,89 @@ impl OnionMessageRecipient {
+}
+
+/// A enum to handle received [`OnionMessage`]
+pub enum ReceivedOnionMessage<R, T>
```
Let's move this above `ResponseFunction` so that the structs can be read from outside-in.

shaavan replied:
Addressed in [2806122](https://github.com/lightningdevkit/rust-lightning/pull/2907/commits/2806122f65d57d3a6a91591aff6e9ccb7590af13)
Thanks!

jkczyz commented on lightning/src/onion_message/messenger.rs:
```diff
@@ -243,6 +244,89 @@ impl OnionMessageRecipient {
+T: OnionMessageContents
+{
+	WithReplyPath {
+		responder: Responder<R, T>,
```
nit: let's make this the last field. Might be some sort of compiler optimizations if the first two fields in each variant are the same.

shaavan replied:
Addressed in [2806122](https://github.com/lightningdevkit/rust-lightning/pull/2907/commits/2806122f65d57d3a6a91591aff6e9ccb7590af13)
Thanks!

jkczyz commented on lightning/src/onion_message/messenger.rs:
```diff
@@ -243,6 +244,89 @@ impl OnionMessageRecipient {
+R: RespondFunction<T>,
+T: OnionMessageContents
+{
+	fn new(messenger_function: R, message: T, reply_path_option: Option<BlindedPath>, path_id: Option<[u8; 32]> ) -> Self {
```
No need to use `option` in the name here, IMO.

shaavan replied:
I get your point!
I still think having a slightly different name could be handy, especially since we're using reply_path: <BlindedPath> in the function.
I've simplified the name a bit to make it less wordy but still meaningful.
Let me know if we want to simplify it further, and I would love to make the change!

jkczyz commented on lightning/src/onion_message/messenger.rs:
```diff
@@ -922,22 +989,36 @@ where
+						// message_type is defined outside of closure to avoid burrowing
+						// issues when using msg for creating ReceivedOnionMessage::New()
+						let message_type = msg.msg_type();
+						let closure = |response, reply_path: BlindedPath| {
+							self.handle_onion_message_response(
+								response, reply_path, format_args!(
+									"when responding to {} onion message with path_id {:02x?}",
+									message_type,
+									path_id
+								)
+							);
+						};
```
I *think* we can DRY this up by moving the `closure` definition outside the `match`.

shaavan replied:
I gave this a shot, but since the type of `response` depends on the type of `message`, creating and using the same function for both handle_message and handle_custom_message, became tricky.
So I have hit a roadblock on this idea for now!

Let me know if there's some simpler way to achieve DRY code here, and I would love to make the change!
Thanks!

jkczyz replied:
Good point! We'll have to define the closure using a local macro definition instead.

jkczyz commented on lightning/src/onion_message/messenger.rs:
```diff
@@ -922,22 +989,36 @@ where
-								path_id
-							)
-						);
+						// message_type is defined outside of closure to avoid burrowing
+						// issues when using msg for creating ReceivedOnionMessage::New()
```
Probably not necessary to explain this.

jkczyz reviewed (commented):
Concept ACK

shaavan commented:
Updated from [pr2907.03](https://github.com/shaavan/rust-lightning/commits/pr2907.03) to [pr2907.04](https://github.com/shaavan/rust-lightning/commits/pr2907.04) ([diff](https://github.com/shaavan/rust-lightning/compare/pr2907.03..pr2907.04)):
Addressed @jkczyz comments

Changes are detailed in the commit message and can be viewed through the [diff](https://github.com/shaavan/rust-lightning/compare/pr2907.03..pr2907.04).

jkczyz commented on lightning/src/onion_message/messenger.rs:
```diff
@@ -243,6 +244,86 @@ impl OnionMessageRecipient {
+	WithReplyPath {
+		message: T,
+		path_id: Option<[u8; 32]>,
+		responder: Responder<R, T>,
```
Hmmm... it's not really a reply but rather a message that can be replied to. Regardless of naming, I'm not sure I see how such an interface would be any better or work for that matter. The handler would need a reference to `OnionMessenger`, which would result in a circular reference since `OnionMessenger` already has a reference to the handler. Am I misunderstanding what you're proposing?

jkczyz replied:
@shaavan Discussed further offline with @TheBlueMatt. Seems we'll run into the circular reference and some other lifetime issues in the async case. We'd soon see this when writing a test where the handler tries to store the `ReceivedOnionMessage` for later processing.

So we'll have to drop the `ResponseFunction` unfortunately. Instead, we'll need to do something like the following:
- Have `Responder::respond` return some sort of `ResponseInstructions<T>` (exact name TBD... maybe `OnionMessageResponse` instead?) which would include the response message and (in a follow-up) a way to indicated that a reply path should be built to include with the response
- Change the handlers back to have a return value only this time using type `ResponseInstructions<T>` where `T` is the concrete message type for the handler.
- Instead of using an `Option` in the handler's return type you could possibly make `ResponseInstructions` an enum with a `NoResponse` variant, FWIW.
- When OnionMessenger gets the returned `ResponseInstructions` it would respond appropriately.

That way a handler can take ownership of `ReceivedOnionMessage` and produce a `ResponseInstructions` asynchronously. In another follow-up PR we would add a method to `OnionMessenger` that takes a `ResponseInstructions` and responds just as it does in the sync case.

shaavan commented:
Updated from [pr2907.05](https://github.com/shaavan/rust-lightning/commits/pr2907.05) to [pr2907.06](https://github.com/shaavan/rust-lightning/commits/pr2907.06) ([diff](https://github.com/shaavan/rust-lightning/compare/pr2907.05..pr2907.06)):
Addressed @jkczyz comment.

Changes:
- Introduce the ResponseInstruction enum. Details on the reasoning, functionality, and approach are documented in [commit message](https://github.com/shaavan/rust-lightning/commit/3dd0fc0af79419aa20606d3784e1e7a38b841628).

shaavan commented:
Updated from [pr2907.06](https://github.com/shaavan/rust-lightning/commits/pr2907.06) to [pr2907.07](https://github.com/shaavan/rust-lightning/commits/pr2907.07) ([diff](https://github.com/shaavan/rust-lightning/compare/pr2907.06..pr2907.07)):
Addressed @jkczyz comments

Updates:
1. Rebase over main to resolve merge conflicts.
2. Addressed generic renaming. [comment](https://github.com/lightningdevkit/rust-lightning/pull/2907#discussion_r1515110182)
3. Introduce code Documentation.

shaavan commented on lightning/src/onion_message/messenger.rs:
```diff
@@ -243,6 +244,90 @@ impl OnionMessageRecipient {
+/// - `WithoutReplyPath`: Represents a message without the possibility of response because of absence of a reply_path.
+///   It contains the original message and an optional path id.
+///
+pub enum ReceivedOnionMessage<T>
```
With the enum architecture, I wanted to ensure a compile-time check for the variant of `ReceivedOnionMesssage`, and I have tried to structure the code in such a way that we are sure at compile time that we are dealing with the `ReceivedOnionMessage::HaveReplyPath` variant and hence _Some_ Responder when we are calling the `respond` function.

Having a simple structure of a `message` and an `Option<Responder>` would have led to a run-time check and not a compile-time check for the Responder.

jkczyz replied:
Discussed offline with @shaavan. The `Option` will work just as fine, so let's go with that approach as suggested. There is an open question about how we should communicate the `path_id`. Another parameter would also work there. But given we aren't using it yet we can punt on that.

shaavan replied:
Addressed in [a595ac6](https://github.com/lightningdevkit/rust-lightning/pull/2907/commits/a595ac6f9c68c7ab3d8650873c2f3039e9092571)

Thanks!

shaavan replied:
Dropped the `ReceivedOnionMessage` in [8fefe50](https://github.com/lightningdevkit/rust-lightning/pull/2907/commits/8fefe50780a5631c409e9039e01a4a3b3e967c1b)
Thanks a lot for the clarification!

jkczyz commented on lightning/src/onion_message/messenger.rs:
```diff
@@ -243,6 +244,90 @@ impl OnionMessageRecipient {
+/// It includes the following variants:
+/// - `HaveResponse`: Indicates that a response is available, and can be sent using reply_path.
+/// - `NoResponse`: Indicates that no response is available.
+pub enum ResponseInstruction<T: OnionMessageContents> {
```
Yeah, we also don't want to give the handler the option to return a response if the sender didn't supply a reply path. Although, in the current formulation that is still possible because `ResponseInstruction::HaveResponse` could in theory be constructed by the handler with an arbitrary `BlindedPath`. So it would be better to have that variant wrap a struct that only `Responder` can construct. Then that struct can be re-used in a third variant (next PR) for signaling a reply path should be included with the response.

> Would this be simpler as a struct (and the handle_*message methods would return an `Option`)? That way `respond` would just return a struct containing a message, path pair and the user can just pass those to send_message.

Hmm... I think we want some sort of instructions as it can include context for `OnionMessenger` to log. Like the corresponding `path_id` that it is responding to, which come to think of it is why we probably want to include that in the responder so it can be passed along to the `ResponseInstruction`. Right now it is fine without it for the sync case but we'll need it for the async case.

shaavan commented:
Updated from [pr2907.07](https://github.com/shaavan/rust-lightning/commits/pr2907.07) to [pr2907.08](https://github.com/shaavan/rust-lightning/commits/pr2907.08) ([diff](https://github.com/shaavan/rust-lightning/compare/pr2907.07..pr2907.08)):
Addressed @TheBlueMatt, and @jkczyz comments:

1. Simplified the ReceivedOnionMessage into a struct, that takes responder as an enum.
2. Removed path_id as a member variable from ReceivedOnionMessage
3. Introduce the OnionMessageResponse struct that allows keeping response and reply_path private while using them along with ResponseInstruction.

jkczyz commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -9427,121 +9428,131 @@ where
-						return Some(OffersMessage::InvoiceError(error.into()));
-					},
-				};
+		if let ReceivedOnionMessage { message, responder: Some(responder) } = message {
```
We still should handle each message even if we don't have a responder. As of now, we don't include a reply path with an invoice. So if we skipped handling such a message we would never pay an invoice from an LDK node.

shaavan replied:
Addressed in [8fefe50](https://github.com/lightningdevkit/rust-lightning/pull/2907/commits/8fefe50780a5631c409e9039e01a4a3b3e967c1b)
Thanks!

shaavan commented:
Updated from [pr2907.08](https://github.com/shaavan/rust-lightning/commits/pr2907.08) to [pr2907.09](https://github.com/shaavan/rust-lightning/commits/pr2907.09) ([diff](https://github.com/shaavan/rust-lightning/compare/pr2907.08..pr2907.09)):
Addressed @jkczyz comments

Updates:

1. Rename the member variable of `OnionMessageResponse` to `message` to better suit its role.
2. Update `ResponseInstructions` to follow the tuple syntax.
3. Update and expand Docs.
4. Drop the `ReceivedOnionMessage` and directly pass the message and Option<Responder> to the handlers.
5. Update `handle_message` and `handle_custom_message` to handle the message even when there’s no responder.

shaavan commented on lightning/src/onion_message/messenger.rs:
```diff
@@ -243,6 +244,77 @@ impl OnionMessageRecipient {
+
+/// This struct contains the information needed to forward a response.
+pub struct OnionMessageResponse<T: OnionMessageContents> {
+	response: T,
```
I think that will be a better name for it!
Addressed in [8fefe50](https://github.com/lightningdevkit/rust-lightning/pull/2907/commits/8fefe50780a5631c409e9039e01a4a3b3e967c1b)
Thanks for the suggestion!

shaavan commented on lightning/src/onion_message/messenger.rs:
```diff
@@ -243,6 +244,77 @@ impl OnionMessageRecipient {
+pub enum ResponseInstruction<T: OnionMessageContents> {
+	HaveResponse {
+		response: OnionMessageResponse<T>
+	},
```
Addressed in [8fefe50](https://github.com/lightningdevkit/rust-lightning/pull/2907/commits/8fefe50780a5631c409e9039e01a4a3b3e967c1b)
Thanks!

shaavan commented:
Updated from [pr2907.09](https://github.com/shaavan/rust-lightning/commits/pr2907.09) to [pr2907.10](https://github.com/shaavan/rust-lightning/commits/pr2907.10) ([diff](https://github.com/shaavan/rust-lightning/compare/pr2907.09..pr2907.10)):
Addressed @jkczyz comments

Updates:
1. Broke the long lines of code to ensure the lines are under 100 characters max.
2. Simplified the responder logic by using a match responder instead of a nested function.

shaavan commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -9427,45 +9428,45 @@ where
+					Err(error) => return handle_error_response(responder, OffersMessage::InvoiceError(error.into())),
 				};
-				let invoice_request = match invoice_request.verify(expanded_key, secp_ctx) {
+				let invoice_request = match invoice_request.clone().verify(expanded_key, secp_ctx) {
```
Yep! I just checked!
Since we are moving the value, and `InvoiceRequest` doesn't implement `Copy`, that's why we are using the `Clone` here!

shaavan replied:
Right! Thanks for pointing that out!
Corrected and removed the redundant clone in [5b05294](https://github.com/lightningdevkit/rust-lightning/commit/5b05294de2454ce817d65334c9afc53f6f844a87)
Thanks!

shaavan commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -9427,45 +9428,45 @@ where
-						let error = Bolt12SemanticError::InvalidMetadata;
-						return Some(OffersMessage::InvoiceError(error.into()));
-					},
+					Err(()) => return handle_error_response(responder, OffersMessage::InvoiceError(Bolt12SemanticError::InvalidMetadata.into())),
```
Makes sense! Sounds like a good idea to maintain code readability!
Updated in [95d67f7](https://github.com/lightningdevkit/rust-lightning/pull/2907/commits/95d67f7b7d9a4ad3731eb7bceb678c11b2e9a964)
Thanks!

shaavan commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -9427,45 +9428,45 @@ where
+			} else {
+				ResponseInstruction::NoResponse
+			}
+		}
```
Thanks for the pointer!
Addressed in [95d67f7](https://github.com/lightningdevkit/rust-lightning/pull/2907/commits/95d67f7b7d9a4ad3731eb7bceb678c11b2e9a964)

shaavan commented:
Updated from [pr2907.10](https://github.com/shaavan/rust-lightning/commits/pr2907.10) to [pr2907.11](https://github.com/shaavan/rust-lightning/commits/pr2907.11):

Changes:
1. Rebase on main.
2. Cleanup commit history.
3. Expand commit message
4. Update PR description.

jkczyz commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -9450,24 +9451,32 @@ where
 		let expanded_key = &self.inbound_payment_key;
 
-		match message {
+		let responder = match responder {
+			Some(responder) => responder,
+			None => return ResponseInstruction::NoResponse,
+		};
```
This needs to go inside the `OffersMessage::InvoiceRequest` arm. Otherwise, we'll never reach the `OffersMessage::Invoice` arm for invoices sent by LDK since we currently don't have a way to include a reply path. Thus, we'll never pay the invoice.

This is only a problem for the offer flow. For the refund flow, we do so in a different manner: https://github.com/lightningdevkit/rust-lightning/blob/b8b1ef3149f26992625a03d45c0307bfad70e8bd/lightning/src/ln/channelmanager.rs#L7993-L7998

shaavan replied:
That makes sense!
I tinkered with your suggestion and I realized I needed to clone the responder to allow it to be used later down the code.

So to avoid this, I have experimented with using a `closure` approach in [51730b4](https://github.com/lightningdevkit/rust-lightning/commit/51730b4286ba568fc7e3de8743efe0804bb0a895)
This solves the problem of not correctly going through the flow, while also avoiding cloning the responder down the line.

Let me know if the approach seems sound!
Thanks a lot, for the suggestions and pointers!

jkczyz replied:
Right, good point! Maybe it would be better to explicitly respond inline throughout instead of having a catch-all at the end of the handler. We already are doing that for early returns. Then you don't need the closure and can use the above code.

For the `OffersMessage::Invoice` arm, you'll need to store the response in variable like we do at the end of the `OffersMessage::InvoiceRequest` arm instead of having the `match` evaluate to the response. Both arms would respond there.

shaavan commented:
Updated from [pr2907.11](https://github.com/shaavan/rust-lightning/commits/pr2907.11) to [pr2907.12](https://github.com/shaavan/rust-lightning/commits/pr2907.12) ([diff](https://github.com/shaavan/rust-lightning/compare/pr2907.11..pr2907.12)):

Updates:
1. Fixed failing test by using closure to correctly handle `responder.response()`.
2. Added doc implementation of `msg_type` to fix a failing test.

shaavan commented:
Updated from [pr2907.12](https://github.com/shaavan/rust-lightning/commits/pr2907.12) to [pr2907.13](https://github.com/shaavan/rust-lightning/commits/pr2907.13) ([diff](https://github.com/shaavan/rust-lightning/compare/pr2907.12..pr2907.13)):

Update:
1. Corrected & expanded `Responder` docs.
2. Fix fuzz testing impls.

shaavan commented:
Updated from [pr2907.14](https://github.com/shaavan/rust-lightning/commits/pr2907.14) to [pr2907.15](https://github.com/shaavan/rust-lightning/commits/pr2907.15) ([diff](https://github.com/shaavan/rust-lightning/compare/pr2907.14..pr2907.15)):
Addressed @TheBlueMatt comments

Updates:
1. Remove redundant referencing in msg_type.
2. Update `Responder` docs to be clearer and true to the code.
3. Spaces -> Tabs
4. Remove redundant docs for private fields.
5. Remove redundant `PhantomData` from Responder, and make `responder.reply_path` public.

shaavan commented on fuzz/src/onion_message.rs:
```diff
@@ -111,6 +111,9 @@ impl OnionMessageContents for TestCustomMessage {
 		CUSTOM_MESSAGE_TYPE
 	}
+	fn msg_type(&self) -> &'static str {
+		&"CustomMessageType"
```
Addressed in [ea40d0c](https://github.com/lightningdevkit/rust-lightning/commit/ea40d0ca0029e68246d8f7afd6ff4998067b2a82)
Thanks!

shaavan commented on lightning/src/onion_message/messenger.rs:
```diff
@@ -243,6 +245,52 @@ impl OnionMessageRecipient {
 	}
 }
 
+/// The `Responder` struct manages [`ResponseInstruction`] for a given response.
```
I intended it to mean something similar, but I think your suggested phrasing will be a much better fit.
I have updated the docs, by taking references from comments.
Let me know if the docs sound more correct now.
Thanks for the suggestion!

shaavan commented on lightning/src/onion_message/messenger.rs:
```diff
@@ -243,6 +245,52 @@ impl OnionMessageRecipient {
+    }
+
+	/// Creates the appropriate [`ResponseInstruction`] for a given response.
+    pub fn respond(self, response: T) -> ResponseInstruction<T>{
```
Corrected in [ea40d0c](https://github.com/lightningdevkit/rust-lightning/commit/ea40d0ca0029e68246d8f7afd6ff4998067b2a82)

Feels like VS code has a will of its own when it comes to Tabs and Spaces! xD

shaavan commented on lightning/src/onion_message/messenger.rs:
```diff
@@ -243,6 +245,52 @@ impl OnionMessageRecipient {
+
+/// This struct contains the information needed to reply to a received message.
+pub struct OnionMessageResponse<T: OnionMessageContents> {
+    /// The response message that has been generated.
```
Thanks for the pointer! I tried to fill in some docs to pass the CI
Glad to know that we don't enforce docs for private fields!
Cleaned up the redundant docs in [ea40d0c](https://github.com/lightningdevkit/rust-lightning/commit/ea40d0ca0029e68246d8f7afd6ff4998067b2a82)
Thanks!

shaavan commented on lightning/src/onion_message/messenger.rs:
```diff
@@ -243,6 +245,52 @@ impl OnionMessageRecipient {
+/// This struct contains the information needed to reply to a received message.
+pub struct OnionMessageResponse<T: OnionMessageContents> {
+    /// The response message that has been generated.
+    message: T,
```
Makes sense!
Make the `responder.reply_path` public in [ea40d0c](https://github.com/lightningdevkit/rust-lightning/commit/ea40d0ca0029e68246d8f7afd6ff4998067b2a82)

jkczyz replied:
Hmm.. no, I think we'd rather have the async version use `Responder` to construct a  `ResponseInstruction` and pass *that* to `OnionMessenger`. This is because it is needed to tell `OnionMessenger` whether or not to create a reply path when sending the response.

The user only needs to know that a reply path exists, which is conveyed through the presence of a `Responder`. There's no need for them to explicitly hold a `BlindedPath`, whether for sending the response or for including in the response's reply path.

The additional method on `OnionMessenger` was to be in the follow-up PR. It would basically be `OnionMessenger::handle_onion_message_response` only we really want the response instructions to convey the context for logging.

jkczyz replied:
> I'm not quite sure I understand how this is related? Whether to include a reply path in a message seems unrelated to whether the message is a response, and the user should be able to specify it independently. I mean I'm not against doing it via the responder, but ISTM its all equivalent modulo logging.

It's useful here for the non-async case, where the handler returns `ResponseInstruction`. I suppose it doesn't matter if `Responder` returns `ResponseInstruction` or if it returns `OnionMessageResponse` and the handler manually wraps it in the desired `ResponseInstruction` variant.

Anyhow, I think my later point still stands that there's no need for the user to touch the `BlindedPath`.

> No reason not to do it now? If its just a new method that logs differently that seems like we could just include it in this PR...

Your call. Either is fine by me but I'd figure we'd try to land this sooner. The logging will require additional context (e.g., path id) in the `Responder` which would get moved to `OnionMessageResponse` like is done with the reply path.

shaavan replied:
I have made the `responder.reply_path` private again in [5b05294](https://github.com/lightningdevkit/rust-lightning/commit/5b05294de2454ce817d65334c9afc53f6f844a87), and I agree it's better to introduce a new function in `OnionMessenger` than to make `reply_path` public
Considering that the new function will be utilized for the async case, and has not been directly used in this PR. I feel like taking it in a follow-up might be better. But let me know whichever works best. Meanwhile, I am gonna start working on this new commit!

jkczyz replied:
> > It's useful here for the non-async case, where the handler returns ResponseInstruction. I suppose it doesn't matter if Responder returns ResponseInstruction or if it returns OnionMessageResponse and the handler manually wraps it in the desired ResponseInstruction variant.
> 
> Still not sure I understand this argument. `ResponseInstruction` is one way to communicate to the `OnionMessenger` whether to include a reply path (well, it will be once it is fully filled out and has a `WithReplyPath` variant), but we could certainly communicate that via a bool in the message sending function as well?

I was referring to the non-async case where we aren't calling a function but rather are returning a value from the handler.

jkczyz commented on lightning/src/onion_message/messenger.rs:
```diff
@@ -243,6 +244,47 @@ impl OnionMessageRecipient {
+	/// Indicates that a response should be sent without including a reply path for the receiver to respond back.
+	WithoutReplyPath(OnionMessageResponse<T>),
+	/// Indicates that there's no response to send back.
+    NoResponse,
```
Tab not spaces.

shaavan replied:
Done! [5b05294](https://github.com/lightningdevkit/rust-lightning/commit/5b05294de2454ce817d65334c9afc53f6f844a87)

jkczyz commented:
> 5. Remove redundant `PhantomData` from Responder, and make `responder.reply_path` public.

I may have missed the discussion, but why was this removed? Didn't we want to enforce that the responder only worked iwth one message type?

jkczyz commented:
> I'm not sure why we would? I mean, its not really a problem to do so, and to be clear we do restrict the ability to respond with an arbitrary message currently as we require that the return type matches in `handle_*_message`, but I'm not entirely sure we need to?

Yeah, spoke with @shaavan  earlier about this. It's definitely not needed in this PR for the point you gave. The question later is if we want to provide the flexibility to respond with any arbitrary message for the async case, or if the `Responder` should prevent the user from doing so. I don't have a strong opinion. Seems it provides some marginal type checking. Presumably the recipient will authenticate that the messages sent over the reply path is for the desired purpose.

shaavan commented:
Updated from [pr2907.15](https://github.com/shaavan/rust-lightning/commits/pr2907.15) to [pr2907.16](https://github.com/shaavan/rust-lightning/commits/pr2907.16) ([diff](https://github.com/shaavan/rust-lightning/compare/pr2907.15..pr2907.16)):
Addressed @TheBlueMatt, and @jkczyz comments

Updates:
1. Remove redundant variable in `handle_custom_message`.
2. Remove closure, take message by value and return inline in `handle_message` function.
3. Fix tabs not spaces and other styling discrepancies.
4. Make `responder.reply_path` private again.

jkczyz commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -9450,24 +9451,32 @@ where
 {
-	fn handle_message(&self, message: OffersMessage) -> Option<OffersMessage> {
+	fn handle_message(&self, message: OffersMessage, responder: Option<Responder>) -> ResponseInstruction<OffersMessage>
+	{
```
Keep the brace on the previous line here, too.

shaavan replied:
Corrected in [77c393a](https://github.com/lightningdevkit/rust-lightning/commit/77c393a6eec06158cbe5fe33608fbe00e2386135)
Thanks!

jkczyz commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -9559,13 +9569,19 @@ where
+				};
+
+				match (responder, response) {
+					(Some(responder), Some(response)) => return responder.respond(response),
```
Let's log if we have a response but no responder.

shaavan replied:
Introduced a `log_trace!` in [77c393a](https://github.com/lightningdevkit/rust-lightning/commit/77c393a6eec06158cbe5fe33608fbe00e2386135)

shaavan commented on lightning/src/onion_message/messenger.rs:
```diff
@@ -243,6 +245,52 @@ impl OnionMessageRecipient {
 
+/// The `Responder` struct manages [`ResponseInstruction`] for a given response.
+pub struct Responder<T: OnionMessageContents>
+{
```
Addressed in [5b05294](https://github.com/lightningdevkit/rust-lightning/commit/5b05294de2454ce817d65334c9afc53f6f844a87)
Thanks!

jkczyz commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -9538,13 +9547,14 @@ where
+						Err(error) => return responder.respond(error),
 					}
-				}
+				};
```
semi-colon not needed

shaavan replied:
Cleaned up in [77c393a](https://github.com/lightningdevkit/rust-lightning/commit/77c393a6eec06158cbe5fe33608fbe00e2386135)
Thanks!

shaavan commented on fuzz/src/onion_message.rs:
```diff
@@ -123,8 +126,12 @@ struct TestCustomMessageHandler {}
-	fn handle_custom_message(&self, _msg: Self::CustomMessage) -> Option<Self::CustomMessage> {
-		Some(TestCustomMessage {})
+	fn handle_custom_message(&self, message: Self::CustomMessage, responder: Option<Responder>) -> ResponseInstruction<Self::CustomMessage> {
+		let response = message;
```
Fixed in [5b05294](https://github.com/lightningdevkit/rust-lightning/commit/5b05294de2454ce817d65334c9afc53f6f844a87)
Thanks!

jkczyz commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -9559,13 +9569,19 @@ where
+				return ResponseInstruction::NoResponse
 			},
-		}
+		};
```
semi-colon not needed

shaavan replied:
Cleaned up in [77c393a](https://github.com/lightningdevkit/rust-lightning/commit/77c393a6eec06158cbe5fe33608fbe00e2386135)
Thanks!

shaavan commented:
Updated from [pr2907.16](https://github.com/shaavan/rust-lightning/commits/pr2907.16) to [pr2907.17](https://github.com/shaavan/rust-lightning/commits/pr2907.17) ([diff](https://github.com/shaavan/rust-lightning/compare/pr2907.16..pr2907.17)):
Addressed @jkczyz comments

1. Fixed styling, and removed redundant `;` in handle_message.
2. Introduce a `log_trace!` for the case when we have a response but no responder.

shaavan commented:
Updated from [pr2907.17](https://github.com/shaavan/rust-lightning/commits/pr2907.17) to [pr2907.18](https://github.com/shaavan/rust-lightning/commits/pr2907.18) ([diff](https://github.com/shaavan/rust-lightning/compare/pr2907.17..pr2907.18)):
Addressed @TheBlueMatt comments

Updates:
1. Removed the unnecessary diff from handle_custom_message
2. Introduced a basic `handle_onion_message_response` wrapper that logs the appropriate message before passing the response to `handle_onion_message_response` for proper sending.

shaavan commented on lightning/src/onion_message/functional_tests.rs:
```diff
@@ -122,15 +125,19 @@ impl Drop for TestCustomMessageHandler {
-			Some(expected_msg) => assert_eq!(expected_msg, msg),
-			None => panic!("Unexpected message: {:?}", msg),
+			Some(expected_msg) => assert_eq!(expected_msg, message),
+			None => panic!("Unexpected message: {:?}", message),
```
Done! [146f0ae](https://github.com/lightningdevkit/rust-lightning/commit/146f0ae369d7fa7006526850e35a45a2015f206c)

shaavan commented:
@TheBlueMatt 

> if we get a PR up and ACK'd tomorrow that exposes a public wrapper around handle_onion_message_response and can land both tomorrow this can slip into the next release

I have added a basic public wrapper as a separate commit in this PR so that reviewers can have an easier time reviewing it in the context of the changes in this PR. Let me know if the wrapper seems sound!
Thanks!

shaavan commented:
Updated from [pr2907.18](https://github.com/shaavan/rust-lightning/commits/pr2907.18) to [pr2907.19](https://github.com/shaavan/rust-lightning/commits/pr2907.19) ([diff](https://github.com/shaavan/rust-lightning/compare/pr2907.18..pr2907.19)):
Addressed @TheBlueMatt comment

Update:
1. Fixed ci failures.

shaavan commented:
Updated from [pr2907.19](https://github.com/shaavan/rust-lightning/commits/pr2907.19) to [pr2907.20](https://github.com/shaavan/rust-lightning/commits/pr2907.20) ([diff](https://github.com/shaavan/rust-lightning/compare/pr2907.19..pr2907.20)):

Changes:
1. Refactor handle_onion_message_response argument to take message_type, and path_id so that they can be used from within the function if needed.
2. Introduce test_coverage for handle_onion_message_response_with_context simulating aync message response.

jkczyz commented on lightning/src/onion_message/messenger.rs:
```diff
@@ -243,6 +243,46 @@ impl OnionMessageRecipient {
 
+
+/// The `Responder` struct creates an appropriate [`ResponseInstruction`]
+/// for a response based on our ability to respond.
```
s/a response based on our ability to respond/responding to a message

jkczyz commented on lightning/src/onion_message/messenger.rs:
```diff
@@ -243,6 +243,46 @@ impl OnionMessageRecipient {
+/// The `Responder` struct creates an appropriate [`ResponseInstruction`]
+/// for a response based on our ability to respond.
+pub struct Responder {
+	/// The path along which the response can be sent.
```
s/the/a

jkczyz commented on lightning/src/onion_message/messenger.rs:
```diff
@@ -243,6 +243,46 @@ impl OnionMessageRecipient {
+}
+
+/// This struct contains the information needed to reply to a received message.
+#[allow(unused)]
```
Is this needed?

shaavan replied:
The commit was triggering some "unused" warnings (and I believe some failures) in commit-wise CI.
So I added this to prevent those ci issues.

jkczyz commented on lightning/src/onion_message/messenger.rs:
```diff
@@ -243,6 +243,46 @@ impl OnionMessageRecipient {
+
+/// `ResponseInstruction` represents instructions for responding to received messages.
+pub enum ResponseInstruction<T: OnionMessageContents> {
+	/// Indicates that a response should be sent without including a reply path for the receiver to respond back.
```
Please wrap at 100 characters.

jkczyz replied:
s/receiver/recipient

jkczyz commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -9450,24 +9451,31 @@ where
-					Err(error) => return Some(OffersMessage::InvoiceError(error.into())),
+					Err(error) => {
+						return responder.respond(OffersMessage::InvoiceError(error.into()));
+					}
```
Add trailing comma. We'll enforce this with `rustfmt` in the future.

jkczyz commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -9450,24 +9451,31 @@ where
 					Err(()) => {
 						let error = Bolt12SemanticError::InvalidMetadata;
-						return Some(OffersMessage::InvoiceError(error.into()));
-					},
```
Please keep commas at the end of match arms. Likewise below. In general, try to keep the patch set limited.

jkczyz commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -9559,11 +9568,24 @@ where
+						);
+						return ResponseInstruction::NoResponse
+					},
+					_ => return ResponseInstruction::NoResponse
```
Trailing comma needed here, too.

jkczyz commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -9559,11 +9568,24 @@ where
+							self.logger,
+							"A response was generated, but there is no reply_path specified for sending the response."
+						);
+						return ResponseInstruction::NoResponse
```
Add sem-colon

jkczyz commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -9559,11 +9568,24 @@ where
 			OffersMessage::InvoiceError(invoice_error) => {
 				log_trace!(self.logger, "Received invoice_error: {}", invoice_error);
-				None
+				return ResponseInstruction::NoResponse
```
Add semi-colon

shaavan replied:
Thanks for all the nit suggestions!
Addressed them in [fa60217](https://github.com/lightningdevkit/rust-lightning/commit/fa602175b41920409b20bb85b4fef9f531889f9c)

jkczyz commented on lightning/src/onion_message/messenger.rs:
```diff
@@ -980,24 +973,25 @@ where
+						let responder = reply_path.map(|path| Responder::new(path));
+						let response_instructions = self.offers_handler.handle_message(msg, responder);
+						self.handle_onion_message_response(response_instructions, format_args!(
+							"when responding to {} onion message with path_id {:02x?}",
+							message_type,
```
If we aren't reusing this string literal, then the new `msg_type` method isn't useful.

shaavan replied:
I am not sure I understand your comment correctly.
message_type is only used once in the handle_message_function currently (that is in `format_args!()`)
In the follow-up (#2996 )I have passed the `message_type` as an input parameter in `handle_onion_message_response`, where it can be reused if necessary.

So it would be great if you could help me understand how should I be reusing `message_type` in this PR. Thanks!

shaavan replied:
Convert `msg_type()` to the associated function as per our offline discussion in [6109cd0](https://github.com/lightningdevkit/rust-lightning/commit/6109cd0c2e32c5e42f6d5bb8a30ebdd2dde8b350)
Thanks!

jkczyz reviewed (commented):
Mostly just nits.

shaavan commented:
Updated from [pr2907.21](https://github.com/shaavan/rust-lightning/commits/pr2907.21) to [pr2907.22](https://github.com/shaavan/rust-lightning/commits/pr2907.22) ([diff](https://github.com/shaavan/rust-lightning/compare/pr2907.21..pr2907.22)):
Addressed @jkczyz comments

1. Addressed nit changes.

shaavan commented:
Updated from [pr2907.22](https://github.com/shaavan/rust-lightning/commits/pr2907.22) to [pr2907.23](https://github.com/shaavan/rust-lightning/commits/pr2907.23) ([diff](https://github.com/shaavan/rust-lightning/compare/pr2907.22..pr2907.23)):

Updates:
1. Converted msg_type functions into associated functions. This simplifies the function usage.
2. Introduce path_id as a field variable for `Responder`. This change allows us to get path_id internally in handle_onion_message_response, simplifying its input parameters.

jkczyz commented on lightning/src/onion_message/messenger.rs:
```diff
@@ -243,6 +243,51 @@ impl OnionMessageRecipient {
+
+impl Responder {
+	/// Creates a new [`Responder`] instance with the provided reply path.
+	pub fn new(reply_path: BlindedPath, path_id: Option<[u8; 32]>) -> Self {
```
We can make this private since `Responder` is supplied to the handler.

shaavan replied:
Sounds good!
Addressed in d0b266c
Thanks!

shaavan commented:
Updated from [pr2907.23](https://github.com/shaavan/rust-lightning/commits/pr2907.23) to [pr2907.24](https://github.com/shaavan/rust-lightning/commits/pr2907.24) ([diff](https://github.com/shaavan/rust-lightning/compare/pr2907.23..pr2907.24)):
Addressed @jkczyz comment

Updates:
1. Rebase on main
2. Made the `Responder` constructor private.

shaavan commented:
> Sadly it needs to wait on the next release, however.

No problemo! :)
Super happy to finally get it done!

Maybe we might also get the follow-up (#2996) in before the next release!

shaavan commented:
Updated from [pr2907.24](https://github.com/shaavan/rust-lightning/commits/pr2907.24) to [pr2907.25](https://github.com/shaavan/rust-lightning/commits/pr2907.25) ([diff](https://github.com/shaavan/rust-lightning/compare/pr2907.24..pr2907.25)):

Changes:
1. Rebase on main.

@TheBlueMatt, a gentle ping! :)
Since #3017 has landed I think we can go forward with landing this one as well!

jkczyz commented on lightning/src/onion_message/packet.rs:
```diff
@@ -135,6 +135,9 @@ impl<T: OnionMessageContents> OnionMessageContents for ParsedOnionMessageContent
 			&ParsedOnionMessageContents::Custom(ref msg) => msg.tlv_type(),
 		}
 	}
+	fn msg_type() -> &'static str {
```
Hmm... this was requested to avoid having users pass a string to `handle_onion_message_response` in the follow-up. See https://github.com/lightningdevkit/rust-lightning/pull/2907#discussion_r1569541966. We really only care about the class of message we are responding to. In `handle_onion_message_response`, we have the response (not the original request), so calling this on the response message passed in wouldn't give a correct logging. That is, for example, we currently log "when responding to Offers onion message" rather than "when responding to InvoiceRequest onion message". And by making this take `&self` it would log "when responding to Invoice onion message", which is not correct.

jkczyz replied:
This assumes the `OffersMessage` implementation changed. Though for custom messages, I can see how a handler may implement more than one class of messages. So I'm not sure how to best handle this. Either in the async case the user passes in the request message's string or we rephrase the logging maybe?

jkczyz replied:
I guess we could hold the request string in `Responder` and move it to the `ResponseInstructions`.That way the user doesn't need to pass it. @shaavan Let's handle this in the follow-up PR.

jkczyz replied:
As mentioned above, the log is stating which message the response is in response to, not the response itself.

jkczyz replied:
The rest of the message says "with path_id {:02x?}", which refers to the message being respond to. Anyhow, you're right that we should change it to `&self`. Doing so would allow implementations to return the specific message type.

shaavan replied:
I think the suggestion sounds reasonable!
Reverted the `msg_type` function to take `self` as a reference in [32a8615](https://github.com/lightningdevkit/rust-lightning/commit/32a8615de62bc63439b529b827e016d5bff48aa0)
Thanks for the pointers!

shaavan commented:
Updated from [pr2907.25](https://github.com/shaavan/rust-lightning/commits/pr2907.25) to [pr2907.26](https://github.com/shaavan/rust-lightning/commits/pr2907.26) ([diff](https://github.com/shaavan/rust-lightning/compare/pr2907.25..pr2907.26)):
Addressed @TheBlueMatt and, @jkczyz comments

Updates:
1. Reverted the `msg_type` function to take reference to the self, allowing it to return specific message type when needed.

jkczyz commented on lightning/src/onion_message/offers.rs:
```diff
@@ -117,6 +118,9 @@ impl OnionMessageContents for OffersMessage {
 		}
 	}
+	fn msg_type(&self) -> &'static str {
+		"Offers"
```
We'll want to make a different one for each variant now.

shaavan replied:
Addressed in [b4204b7](https://github.com/lightningdevkit/rust-lightning/commit/b4204b77c2417a5ca337dc2c336ae5d0871d421e)
Thanks for the pointer!

jkczyz commented on lightning/src/onion_message/messenger.rs:
```diff
@@ -933,19 +978,18 @@ where
+			let _ = self.find_path_and_enqueue_onion_message(
+				response.message, Destination::BlindedPath(response.reply_path), None,
+				format_args!(
+					"when responding to {} onion message with path_id {:02x?}",
```
How about `"when responding with {} to an onion message with path_id {:02x}"`? I think this would read better given the contexts in which the string will be used.

shaavan replied:
I think the second phrasing suits the context better since it emphasizes the `type` of response.

Updated in [b4204b7](https://github.com/lightningdevkit/rust-lightning/commit/b4204b77c2417a5ca337dc2c336ae5d0871d421e)
Thanks a lot for the suggestion!

shaavan commented:
Updated from [pr2907.26](https://github.com/shaavan/rust-lightning/commits/pr2907.26) to [pr2907.27](https://github.com/shaavan/rust-lightning/commits/pr2907.27) ([diff](https://github.com/shaavan/rust-lightning/compare/pr2907.26..pr2907.27)):
Addressed @TheBlueMatt and, @jkczyz comments

Updates:
1. Introduce a specific msg_type string for each variant of OffersMessage.
2. Update the log_suffix message in handle_onion_message_response.

shaavan commented:
Updated from [pr2907.27](https://github.com/shaavan/rust-lightning/commits/pr2907.27) to [pr2907.28](https://github.com/shaavan/rust-lightning/commits/pr2907.28) ([diff](https://github.com/shaavan/rust-lightning/compare/pr2907.27..pr2907.28)):

Updates:
1. Update the returned string for the `msg_type` function for Custom Message, to match the updated log_suffix better.
2. Fix fuzz test.
